dialogue-core/src/main/java/com/palantir/dialogue/core/NeverThrowChannel.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The contract of {@link Channel} requires that the {@link Channel#execute} method never throws. This is a defensive
+ * backstop so that callers can rely on this invariant.
+ */
+final class NeverThrowChannel implements Channel {
+
+    private static final Logger log = LoggerFactory.getLogger(NeverThrowChannel.class);
+    private final Channel delegate;
+
+    NeverThrowChannel(Channel delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        try {
+            return delegate.execute(endpoint, request);
+        } catch (RuntimeException e) {
```

Comment:
I could go either way on handling Error as well, OutOfMemoryError isn't terribly uncommon, and throwing from unexpected calls when things go poorly could result in additional bugs being surfaced (resource leaks). Thoughts?

Reply:
I think we set [CrashOnOutOfMemoryError](https://github.com/palantir/sls-packaging/blob/07fbbe8e57eb59771d29eff0fda1423d33ec7829/gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.java#L57) but I think isolating NoSuchMethodError so that it only breaks specific codepaths is probably a good idea.

Reply:
CrashOnOutOfMemoryError doesn't crash when you run out of direct memory or (I might be mis-remembering this one) attempt to create an array that's larger than the maximum length. I think ByteArrayOutputStream explicitly thows OME in some cases which doesn't trigger a crash.